### PR TITLE
chapter 7: Sign Up

### PR DESCRIPTION
**Pull request by puteriffah**

### CODE REVIEW

**Comment by puteriffah** [comment]
ready

File: app/assets/stylesheets/custom.scss [hunk 1]
```diff
@@ -32,3 +34,63 @@ textarea {
...
+      border: 0;
+      padding-top: 0;
+    }
+    span {
+      display: block;
+      margin-bottom: 3px;
+      line-height: 1;
+    }
+    h1 {
+      font-size: 1.4em;
```
Comment (saobv-2223) [inline app/assets/stylesheets/custom.scss]: every block should have an empty line among them

File: app/assets/stylesheets/users.scss [hunk 2]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the Users controller here.
+// They will automatically be included in application.css.
+// You can use Sass (SCSS) here: https://sass-lang.com/
```
Comment (saobv-2223) [inline app/assets/stylesheets/users.scss]: remove file that never used

File: app/controllers/users_controller.rb [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+  def show
+    @user = User.find_by id: params[:id]
+  end
+
+  def create
+    @user = User.new user_params
+
+    if @user.save
+      flash[:success] = "Welcome to the Sample App!"
+
```
Comment (saobv-2223) [inline app/controllers/users_controller.rb]:
```suggestion
```

File: app/controllers/users_controller.rb [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+      redirect_to @user
+    else
+      render :new
+    end
```
Comment (saobv-2223) [inline app/controllers/users_controller.rb]: add a flash message for this case

File: app/helpers/users_helper.rb [hunk 5]
```diff
@@ -0,0 +1,7 @@
+module UsersHelper
+    def gravatar_for user
+        gravatar_id = Digest::MD5::hexdigest user.email.downcase
+        gravatar_url = "https://secure.gravatar.com/avatar/#{gravatar_id}"
+        image_tag gravatar_url, alt: user.name, class: "gravatar"
+        end
+end
```
Comment (saobv-2223) [inline app/helpers/users_helper.rb]: indenting only 2 spaces, please

File: app/models/user.rb [hunk 6]
```diff
@@ -1,4 +1,7 @@
 class User < ApplicationRecord
+
```
Comment (saobv-2223) [inline app/models/user.rb]:
```suggestion
```

File: app/views/users/new.html.erb [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+    <%= form_for @user, url: signup_path do |f| %>
+      <% if @user.errors.any? %>
+        <div id="error_explanation">
+          <div class="alert alert-danger">
+            <%= t "the_form_contains" %>
+            <%= pluralize @user.errors.count, "error" %>.
+          </div>
+          <ul>
+          <% @user.errors.full_messages.each do |msg| %>
+          <li><%= msg %></li>
```
Comment (saobv-2223) [inline app/views/users/new.html.erb]:
```suggestion
            <li><%= msg %></li>
```

File: app/views/users/new.html.erb [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+          <% end %>
+          </ul>
+        </div>
+        <% end %>
```
Comment (saobv-2223) [inline app/views/users/new.html.erb]:
```suggestion
      <% end %>
```

File: app/views/users/show.html.erb [hunk 9]
```diff
@@ -0,0 +1,11 @@
+<% provide :title, @user.name %>
+<div class="row">
+<aside class="col-md-4">
+<section class="user_info">
+<h1>
+  <%= gravatar_for @user %>
+  <%= @user.name %>
+</h1>
+</section>
+</aside>
```
Comment (saobv-2223) [inline app/views/users/show.html.erb]: indenting 2 spaces, please

File: config/application.rb [hunk 10]
```diff
@@ -10,4 +10,4 @@ class Application < Rails::Application
     config.i18n.available_locales = [:en, :ru]
     config.i18n.default_locale = :en
   end
-end
+end
```
Comment (saobv-2223) [inline config/application.rb]: add end line, please

**Comment by puteriffah** [comment]
ready

File: app/models/user.rb [hunk 11]
```diff
@@ -19,4 +21,5 @@ class User < ApplicationRecord
   def downcase_email
     email.downcase!
   end
+  before_save { email.downcase! }
```
Comment (saobv-2223) [inline app/models/user.rb]:
```suggestion
  before_save :downcase_email
```and move it above downcase_email function

File: app/views/layouts/application.html.erb [hunk 12]
```diff
@@ -11,10 +11,11 @@
 
   <body>
   <%= render "layouts/header"%>
-
   <div class="container">
-    <%= yield %>
+    <% flash.each do |message_type, message| %>
+    <div class="alert alert-<%= message_type %>"><%= message %></div>
+    <% end %>
+  <%= yield %>
```
Comment (saobv-2223) [inline app/views/layouts/application.html.erb]:
```suggestion
    <div class="alert alert-<%= message_type %>"><%= message %></div>
      <% end %>
    <%= yield %>
```

File: app/views/users/new.html.erb [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+    <%= f.label :name %>
+    <%= f.text_field :name %>
+    <%= f.label :email %>
+    <%= f.email_field :email %>
+    <%= f.label :password %>
+    <%= f.password_field :password %>
+    <%= f.label :password_confirmation, "confirmation" %>
+    <%= f.password_field :password_confirmation %>
+    <%= f.submit "create my account!", class: "btn btn-primary" %>
+    <% end %>
```
Comment (saobv-2223) [inline app/views/users/new.html.erb]: indenting 2 spaces

**Comment by puteriffah** [comment]
ready

File: app/views/layouts/application.html.erb [hunk 14]
```diff
@@ -11,10 +11,11 @@
 
   <body>
   <%= render "layouts/header"%>
-
   <div class="container">
+    <% flash.each do |message_type, message| %>
+    <div class="alert alert-<%= message_type %>"><%= message %></div>
+      <% end %>
```
Comment (saobv-2223) [inline app/views/layouts/application.html.erb]:
```suggestion
    <% flash.each do |message_type, message| %>
      <div class="alert alert-<%= message_type %>"><%= message %></div>
    <% end %>
```

File: app/views/users/new.html.erb [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+      <% if @user.errors.any? %>
+        <div id="error_explanation">
+          <div class="alert alert-danger">
+            <%= t "the_form_contains" %>
+            <%= pluralize @user.errors.count, "error" %>.
+          </div>
+          <ul>
+          <% @user.errors.full_messages.each do |msg| %>
+            <li><%= msg %></li>
+            <% end %>
```
Comment (saobv-2223) [inline app/views/users/new.html.erb]:
```suggestion
          <% end %>
```

File: config/routes.rb [hunk 16]
```diff
@@ -1,6 +1,11 @@
 Rails.application.routes.draw do
+
```
Comment (saobv-2223) [inline config/routes.rb]:
```suggestion
```

**Comment by puteriffah** [comment]
ready

**Review by saobv-2223 — COMMENTED** [review]
merge

**Pull request merged**